src/QsCompiler/CompilationManager/DataStructures.cs

```diff
@@ -41,69 +50,374 @@ internal class CodeLine
+                switch (stringContext)
+                {
+                    case StringContext.NoOpenString:
+                        // Find the next "
+                        index = text.IndexOf('"');
+                        break;
+                    case StringContext.OpenInterpolatedArgument:
+                        // Find the next " or }
+                        index = text.IndexOfAny(new[] { '"', '}' });
+                        break;
+                    case StringContext.OpenInterpolatedString:
+                        // Find the next " or {, neither or which being preceded by \
+                        index = text.IndexOfAny(new[] { '"', '{' });
+                        while (index > 0 && text[index - 1] == '\\')
+                        {
+                            var next = text.Substring(index + 1).IndexOfAny(new[] { '"', '{' });
+                            index = next < 0 ? next : index + 1 + next;
+                        }
+                        break;
+                    case StringContext.OpenString:
+                    case StringContext.OpenStringInOpenInterpolatedArgument:
+                        // Find the next " not preceded by \
+                        index = text.IndexOf('"');
+                        while (index > 0 && text[index - 1] == '\\')
+                        {
+                            var next = text.Substring(index + 1).IndexOf('"');
+                            index = next < 0 ? next : index + 1 + next;
+                        }
+                        break;
+                }
+
+                if (index < 0 || (commentIndex > -1 && commentIndex < index))
+                {
+                    break;
+                }
+
+                var inputDelimiter = text[index].ToString();
+                // If the input is a " preceded by a $ than the input needs to be updated to $"
+                if (index > 0 && inputDelimiter == "\"" && text[index - 1] == '$')
+                {
+                    inputDelimiter = "$\"";
+                }
+
+                var pos = index + stringLength - text.Length;
+                bool validDelimiter;
+                (stringContext, validDelimiter) = MoveToNextState(stringContext, inputDelimiter);
+                if (!validDelimiter)
+                {
+                    // If the erroneous delimiter is $", mark the $ as the error and treat the " as a normal string delimiter
+                    if (inputDelimiter == "$\"")
+                    {
+                        errorDelimiterBuilder.Add(pos - 1); // Backup to the $ character
+                        delimiterBuilder.Add(pos);
+                    }
+                    else
+                    {
+                        errorDelimiterBuilder.Add(pos);
+                    }
+                }
+                else
+                {
+                    delimiterBuilder.Add(pos);
+                }
+                text = text.Substring(index + 1);
```

Comment:
Repeatedly reallocating string consumes excessive memory (strings are immutable). All indexof functions take starting posistion for search, so you don't need to reallocate.

Reply:
I've reworked the function so that it will keep the same text but update the position in that text, instead of allocating new strings for each pass as you pointed out.